gtsfm/utils/io.py

```diff
@@ -180,13 +260,62 @@ def write_images(gtsfm_data: GtsfmData, save_dir: str) -> None:
 
         for i in gtsfm_data.get_valid_camera_indices():
             camera = gtsfm_data.get_camera(i)
-            wRi_quaternion = camera.pose().rotation().quaternion()
-            wti = camera.pose().translation()
-            tx, ty, tz = wti
-            qw, qx, qy, qz = wRi_quaternion
+            # COLMAP exports camera extrinsics (cTw), not the poses (wTc), so must invert
+            iTw = camera.pose().inverse()
+            iRw_quaternion = iTw.rotation().quaternion()
+            itw = iTw.translation()
+            tx, ty, tz = itw
+            qw, qx, qy, qz = iRw_quaternion
 
             f.write(f"{i} {qw} {qx} {qy} {qz} {tx} {ty} {tz} {i} {img_fname}\n")
             # TODO: write out the points2d
+            f.write("TODO\n")
+
+
+def read_points_txt(fpath: str) -> Tuple[Optional[np.ndarray], Optional[np.ndarray]]:
+    """Read 3d points and their associated colors from a COLMAP points.txt file.
+
+    Reference: https://colmap.github.io/format.html#points3d-txt
+
+    Args:
+        fpath: absolute file path to points.txt file
+
+    Returns:
+        point_cloud: float array of shape (N,3)
+        rgb: uint8 array of shape (N,3)
+    """
+    if not Path(fpath).exists():
+        logger.info("%s does not exist", fpath)
+        return None, None
+
+    with open(fpath, "r") as f:
+        data = f.readlines()
+
+    rgb = []
+    point_cloud = []
+    # first 3 lines are information about the file format
+    # line at index 2 will be of the form
+    # "# Number of points: 2122, mean track length: 2.8449575871819039"
+    points_metadata = data[2]
+    j = points_metadata.find(":")
+    k = points_metadata.find(",")
+    expected_num_pts = int(points_metadata[j + 1 : k])
+
+    data = data[3:]
+    for line in data:
+        entries = line.split()
+        x, y, z, r, g, b = entries[1:7]
+
+        point = [float(x), float(y), float(z)]
+        point_cloud += [point]
+        rgb += [(int(r), int(g), int(b))]
+
+    point_cloud = np.array(point_cloud)
+    rgb = np.array(rgb).astype(np.uint8)
+
+    assert point_cloud.shape[0] == expected_num_pts
+    assert rgb.shape[0] == expected_num_pts
```

Comment:
should we remove it or keep it?

Reply:
What would you prefer? Raise a RuntimeError?